README.md
# IstioDemo
Isto demo in a kind-cluster inside an Azure-VM

# Istio Lab Setup

[![Asciinema - Istio Lab Setup](https://asciinema.org/a/481272.svg)](https://asciinema.org/a/481272?autoplay=1)

## Create, configure and login into new azure VM
```
az login

git clone https://github.com/zipizap/IstioDemo.git
cd IstioDemo


# Delete (if exists) RG "rg-istiodemo", and then re-create it from scratch containing a vnet+subnet+nsg and VM where cloudinit will install docker,kubectl,helm,k9s and kind
# The VM is left ready for a user to login via ssh, and use "kind" to create a new kubernetes cluster over docker
./dep.destroyRg_and_startFromScratch.sh


# Ssh-login into VM 
./pdep.ssh.vm.sh
```


## Create kubernetes cluster:

Inside the VM

```
# Assure cloud-init finishes executing and installs kind (takes 5min after VM is booted)
while ! which kind; do echo "waiting cloudinit to finish executing..."; sleep 1; done

# Create kubernetes cluster with kind (over docker) and metallb
kind create cluster
kubectl get pod -A

kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.12.1/manifests/namespace.yaml
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.12.1/manifests/metallb.yaml
LB_NET_PREFIX=$(docker network inspect -f '{{range .IPAM.Config }}{{ .Gateway }}{{end}}' kind | cut -d '.' -f1,2)
LB_NET_IP_FIRST=${LB_NET_PREFIX}.255.200
LB_NET_IP_LAST=${LB_NET_PREFIX}.255.250
kubectl apply -f - <<EOT
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: metallb-system
  name: config
data:
  config: |
    address-pools:
    - name: default
      protocol: layer2
      addresses:
      - ${LB_NET_IP_FIRST}-${LB_NET_IP_LAST}
EOT



```

## Install istio and addons

```
curl -L https://istio.io/downloadIstio | sh -
cd istio-1.13.2
export PATH=$PWD/bin:$PATH
istioctl install --set profile=demo -y

# Enable namespace-auto-injection for default ns
kubectl label namespace default  istio-injection=enabled
kubectl get namespaces --label-columns=istio-injection

# Install addons: kiali, prometheus, etc
kubectl apply -f samples/addons
kubectl rollout status deployment/kiali -n istio-system
ISTIOINGRESSGW_LB_IP=$(kubectl -n istio-system get svc/istio-ingressgateway -o=jsonpath='{.status.loadBalancer.ingress[0].ip}')
echo "$ISTIOINGRESSGW_LB_IP istioigw" | sudo tee -a /etc/hosts
```

## Install an app prepared for istio

```
kubectl apply -f samples/bookinfo/platform/kube/bookinfo.yaml

# Check app services, and pods with auto-injected sidecar 2/2
kubectl get services
timeout 60 watch kubectl get pods
#kubectl describe pod pod-with-sidecar

# Check: pod can connect to the app via service productpage:9080
kubectl exec "$(kubectl get pod -l app=ratings -o jsonpath='{.items[0].metadata.name}')" -c ratings -- curl -sS productpage:9080/productpage | grep -o "<title>.*</title>"

# Expose app via Gateway + VirtualService
kubectl apply -f samples/bookinfo/networking/bookinfo-gateway.yaml
kubectl get service,gateway,virtualservice

# Check: from outside the cluster (from VM), we can connect to the app http://...../productpage
# via istio-ingressgateway > Gateway > VirtualService > Service > Deployment > Pod
curl -s http://istioigw/productpage | grep -o "<title>.*</title>"

# Take note of the following ip: <istio-ingressgateway_ExternalIp>, should be 172.18.255.200
getent hosts istioigw

```

- In your laptop ssh-session, add a Local-port-forward-tunnel towards the <istio-ingressgateway_ExternalIp>:80, should be 172.18.255.200:80 like so
  - source-port:          20080
  - destination ip:port   172.18.255.200:80

And then in your laptop open chrome tab to the application webpage http://127.0.0.1:20080/productpage
and the web-application should load in the laptop




# Istio demo

## Generate traffic
- In VM, leave shell open running:
```
while :; do echo -n "$(date)  "; curl -sv http://istioigw/productpage  2>&1 | grep '< HTTP' ; sleep 0.5; done
```

## Open Istio dashboards

### Kiali
- In VM, leave a shell open running: `kubectl proxy`
- In laptop ssh-session, add Local-port-forwarding-tunnel:
  - source-port:          8001
  - destination ip:port   127.0.0.1:8001
- In laptop, open chrome tab to http://localhost:8001/api/v1/namespaces/istio-system/services/http:kiali:20001/proxy/


### Grafana
- In VM, leave a shell open running: `kubectl -n istio-system port-forward service/grafana 8002:3000`
- In laptop ssh-session, add Local-port-forwarding-tunnel:
  - source-port:          8002
  - destination ip:port   127.0.0.1:8002
- In laptop, open chrome tab to http://localhost:8002/d/G8wLrJIZk/istio-mesh-dashboard?orgId=1&refresh=5s&from=now-1h&to=now

### Jaeger
- In laptop, open chrome tab to http://localhost:8001/api/v1/namespaces/istio-system/services/http:tracing:80/proxy/jaeger/search



## Understand core resources and concepts

- Understand sidecar auto-injection via namespace label (and exclusion)
  - New pods get sidecar
  - Existing pods get sidecar on restart

```
kubectl get namespaces --label-columns=istio-injection
kubectl label namespace default  istio-injection=enabled
```

- Understand Gateway and VirtualService, etc

```
#
# 1 nginx-controller + N Ingress + M Services = 1 Gateway + N VirtualService + M Services
# https://rinormaloku.com/istio-practice-routing-virtualservices/
#
#   . ingressgateway  -  Gateway - VirtualService - [DestinationRule: subsets]     - Services - Deployment  -  Pod
#     lb: external-ip
#                        tcp/80
#                        hosts     hosts
#                                  paths
#                                 [weights/subsets]
#                                                  [subsets (host:mySvc;version:v1]                             app: myApp
#                                                                                                               [version: v1/2]
#
#    *virtualservices* glues *gateway* with *services*
#    *destinationrule* defines *subsets* (from pod-label version:) which can be used by virtualServices
#
# Resume gw, virtualservice, destinationRule: 
# https://medium.com/google-cloud/istio-routing-basics-14feab3c040e
```


- Understand app resources

```
istio-ingressgateway (lb) 

Gateway:  bookinfo-gateway 
  - host, port 

VirtualService:  bookinfo
  - host, path , [weight/subset]

[DestinationRule: bookinfo  (declare subsets v1/v2/v3) ]
  - subsets, host, version

Service: productpage, reviews, details, rating  

Deployment: productpage-v1, reviews-v1, reviews-v2, reviews-v3, details-v1, ratings-v1
```

- Show Traffic shifting: Weight-based routing
  - Add DestinationRules to define subsets, and add VirtualService to route traffic based on weights to each subset
  - Wait 2min to let traffic-average to converge
  - Kiali: review traffic distribution % on service "review"
```
kubectl apply -f - <<EOT
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: reviews
spec:
  host: reviews
  subsets:
  - name: v1
    labels:
      version: v1
  - name: v2
    labels:
      version: v2
  - name: v3
    labels:
      version: v3
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: reviews
spec:
  hosts:
    - reviews
  http:
  - route:
    - destination:
        host: reviews
        subset: v1
      weight: 60
    - destination:
        host: reviews
        subset: v2
      weight: 30
    - destination:
        host: reviews
        subset: v3
      weight: 10
EOT
```


- Tour Kiali dashboard

- Tour Grafana dashboard (understand underlying prometheus)

- Tour Jaeger distributed-tracing

- Understand mTLS (intra-mesh, can be mandatory)


# Istio Lab Teardown (destroy VM)

In the same github-repo folder where the previous terraform commands were executed before, run:

```
./terraform_do.sh destroy
```

This deletes the RG, VM and all other azure-resources created by terraform for this demo


mod-vm/Changelog.md
# v0.0.2
+ var `bootfile` (ex: `bootfile="./boot.sh"`) to run sh script at boot (via cloudinit) 

# v0.0.3
+ unique names for resources in same rg (avoid clashed when many vms in same rg)

# v0.0.4
+ Dynamic or Static interal ips
+ os_disk distinct names
+ "cloudinitfile" var

# v0.0.5
+ values.size with default "Standard_B1ls"


cloudinits/client/gen_cloud-config.sh
#!/usr/bin/env bash
# Paulo Aleixo Campos
__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
function shw_info { echo -e '\033[1;34m'"$1"'\033[0m'; }
function error { echo "ERROR in ${1}"; exit 99; }
trap 'error $LINENO' ERR
PS4='████████████████████████${BASH_SOURCE}@${FUNCNAME[0]:-}[${LINENO}]>  '
set -o errexit
set -o pipefail
set -o nounset
set -o xtrace

LxdVmName=ciDnsAuth1

cd "${__dir}"
cat > cloudinit.userdata.yaml <<EOT
#cloud-config

## Order of execution (cat /etc/cloud/cloud.cfg)
## 1) write_files  
## (before packages are installed and bind-user created and package-dirs created with proper permitions...)
write_files:
- path: /var/lib/cloud/scripts/per-once/01.scripts-per-once.sh
  owner: root:root
  permissions: '0544'
  encoding: gz
  content: !!binary |
$(cat ./01.scripts-per-once.sh | gzip | base64 | sed 's/^/    /g')

## 3) packages
package_upgrade: true
packages:
- dnsutils

## 4) scripts-per-once
## ie, scripts in /var/lib/cloud/scripts/per-once 
## as our crafted /var/lib/cloud/scripts/per-once/01.scripts-per-once.sh
## :)

EOT
cat cloudinit.userdata.yaml






cloudinits/client/01.scripts-per-once.sh
#!/usr/bin/env bash
# Paulo Aleixo Campos
__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
function shw_info { echo -e '\033[1;34m'"$1"'\033[0m'; }
function error { echo "ERROR in ${1}"; exit 99; }
trap 'error $LINENO' ERR
PS4='████████████████████████${BASH_SOURCE}@${FUNCNAME[0]:-}[${LINENO}]>  '
exec > >(tee -i /var/tmp/$(date +%Y%m%d%H%M%S.%N)__$(basename $1).log ) 2>&1
set -o errexit
set -o pipefail
set -o nounset
set -o xtrace

apt-get update
apt-get -y install \
    apt-transport-https \
    ca-certificates \
    curl \
    gnupg \
    lsb-release \
    tmux

curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg

echo \
"deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
$(lsb_release -cs) stable" | tee /etc/apt/sources.list.d/docker.list > /dev/null

apt-get update
apt-get -y install docker-ce docker-ce-cli containerd.io
sudo usermod -aG docker uzer

snap install kubectl --classic
snap install helm --classic
curl -Ls https://api.github.com/repos/derailed/k9s/releases/latest | grep -wo "https.*k9s_Linux_x86_64.tar.gz" | xargs curl -sL | tar xvz k9s

curl -Lo ./kind "https://kind.sigs.k8s.io/dl/v0.12.0/kind-$(uname)-amd64"
chmod +x ./kind
mv -v ./kind /usr/local/bin


terraform_do.sh
#!/usr/bin/env bash
# Paulo Aleixo Campos
__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
function shw_info { echo -e '\033[1;34m'"$1"'\033[0m'; }
function error { echo "ERROR in ${1}"; exit 99; }
trap 'error $LINENO' ERR
exec > >(tee -i /tmp/$(date +%Y%m%d%H%M%S.%N)__$(basename $0)_${1:-}.log ) 2>&1
PS4='████████████████████████${BASH_SOURCE}@${FUNCNAME[0]:-}[${LINENO}]>  '
set -o errexit
set -o pipefail
set -o nounset
set -o xtrace

load_gitignoreme.Backend.Provider.source() {
  [[ -r ./gitignoreme.Backend.Provider.source ]] || { echo "Missing file gitignoreme.Backend.Provider.source - aborting!"; exit 1; }
  set +x
  source ./gitignoreme.Backend.Provider.source
  set -x
  terraform version
}

terraform_init() {
  load_gitignoreme.Backend.Provider.source

  if [[ "$BACKEND_PROVIDER" == "RemoteBackend_ProviderViaSPN" ]]; then
    terraform init \
      -input=false \
      -backend-config=resource_group_name=$AZURERM_Backend_RG_NAME \
      -backend-config=tenant_id=$AZURERM_Backend_TENANT_ID \
      -backend-config=subscription_id=$AZURERM_Backend_SUBSCRIPTION_ID \
      -backend-config=client_id=$AZURERM_Backend_CLIENT_ID \
      -backend-config=client_secret="$AZURERM_Backend_CLIENT_SECRET" \
      -backend-config=storage_account_name=$AZURERM_Backend_STA_NAME \
      -backend-config=container_name=$AZURERM_Backend_STA_CONTAINER_NAME \
      -backend-config=key=$AZURERM_Backend_STA_CONTAINER_KEY_NAME
  elif [[ "$BACKEND_PROVIDER" == "LocalBackend_ProviderAzLogin" ]]; then
    terraform init \
      -input=false
  fi
}
terraform_reinit() {
  load_gitignoreme.Backend.Provider.source
  # force cleanup  (-reconfigure should do this, but it seemes not quite in fact... this is easier...)
  [[ -d ./.terraform ]] && rm -rf ./.terraform
  rm -rfv ./.terraform .terraform.lock.hcl *.tfstate *.tfstate.backup "${THE_TFPLAN_FILE}" || true
  terraform_init
}

terraform_plan() {
  load_gitignoreme.Backend.Provider.source
  THE_TFVARS_FILE="${THE_TFVARS_FILE?missing env-var}"
  THE_TFPLAN_FILE="${THE_TFPLAN_FILE?missing env-var}"
  terraform validate \
     -no-color
  [[ -r "${THE_TFPLAN_FILE}" ]] && rm -f "${THE_TFPLAN_FILE}" &>/dev/null
  terraform plan \
     -input=false \
     -var-file=$THE_TFVARS_FILE \
     -out="${THE_TFPLAN_FILE}"
}


terraform_apply() {
  load_gitignoreme.Backend.Provider.source
  terraform apply \
     -input=false \
     "${THE_TFPLAN_FILE}"
}

terraform_destroy_plan() {
  load_gitignoreme.Backend.Provider.source
  THE_TFVARS_FILE="${THE_TFVARS_FILE?missing env-var}"
  terraform plan -destroy \
     -input=false \
     -var-file=$THE_TFVARS_FILE
}

terraform_destroy() {
  load_gitignoreme.Backend.Provider.source
  THE_TFVARS_FILE="${THE_TFVARS_FILE?missing env-var}"
  terraform destroy \
     -input=false -auto-approve \
     -var-file=$THE_TFVARS_FILE
}

show_args_and_exit() {
  cat <<EOT
$(basename $0)
    init|reinit
    plan
    apply
    destroy_plan|destroy
EOT
  exit 1
}

main() {
  [[ "$#" -eq 0 ]] && show_args_and_exit
  while test $# -gt 0
  do
    case "$1" in
      init)
        terraform_init
        ;;
      reinit)
        terraform_reinit
        ;;
      plan)
        terraform_plan
        ;;
      apply)
        terraform_apply
        ;;
      destroy_plan)
        terraform_destroy_plan
        ;;
      destroy)
        terraform_destroy
        ;;
      *)
        show_args_and_exit
        ;;
    esac
    shift
  done
  shw_info "=== Completed successfully ==="
}
main "$@"

